app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt

```diff
@@ -106,65 +123,157 @@ class ImageLoader @Inject constructor(
                 return@launch
             }
 
+            val existsInNotForUploadTable = notForUploadStatusDao.find(imageSHA1)
+
             if (result in arrayOf(Result.NOTFOUND, Result.INVALID) && sha1.isNotEmpty()) {
-                // Query original image.
-                result = querySHA1(imageSHA1)
+                when {
+                    mapResult[imageSHA1] == null -> {
+                        // Query original image.
+                        result = checkWhetherFileExistsOnCommonsUsingSHA1(
+                            imageSHA1,
+                            ioDispatcher,
+                            mediaClient
+                        )
+                        when (result) {
+                            is Result.TRUE -> {
+                                mapResult[imageSHA1] = Result.TRUE
+                            }
+                        }
+                    }
+                    else -> {
+                        result = mapResult[imageSHA1]!!
+                    }
+                }
                 if (result is Result.TRUE) {
                     // Original image found.
                     insertIntoUploaded(imageSHA1, sha1, result is Result.TRUE, false)
                 } else {
-                    // Original image not found, query modified image.
-                    result = querySHA1(sha1)
+                    when {
+                        mapResult[sha1] == null -> {
+                            // Original image not found, query modified image.
+                            result = checkWhetherFileExistsOnCommonsUsingSHA1(
+                                sha1,
+                                ioDispatcher,
+                                mediaClient
+                            )
+                            when (result) {
+                                is Result.TRUE -> {
+                                    mapResult[sha1] = Result.TRUE
+                                }
+                            }
+                        }
+                        else -> {
+                            result = mapResult[sha1]!!
+                        }
+                    }
                     if (result != Result.ERROR) {
                         insertIntoUploaded(imageSHA1, sha1, false, result is Result.TRUE)
                     }
                 }
             }
-            if(mapHolderImage[holder] == image) {
-                if (result is Result.TRUE) holder.itemUploaded() else holder.itemNotUploaded()
+
+            val sharedPreferences: SharedPreferences =
+                context
+                    .getSharedPreferences(ImageHelper.CUSTOM_SELECTOR_PREFERENCE_KEY, 0)
+            val showAlreadyActionedImages =
+                sharedPreferences.getBoolean(
+                    ImageHelper.SHOW_ALREADY_ACTIONED_IMAGES_PREFERENCE_KEY,
+                    true
+                )
+
+            if (mapHolderImage[holder] == image) {
+                if (result is Result.TRUE) {
```

Comment:
U do not need nested conditions here.

Reply:
Can you elaborate on this? Are you trying to say I should combine those nested if/else statements by using &?

Reply:
yes.